pycode/memilio-generation/memilio/generation/scanner.py

```diff
@@ -73,23 +80,80 @@ def create_ast(self: Self) -> None:
-                        argument != "--driver-mode=g++" and argument != "-O3"):
+                        argument != "--driver-mode=g++" and argument != "-O3" and argument != "-Werror" and argument != "-Wshadow"):
                     file_args.append(argument)
         file_args = file_args[1:-4]
+
+        # Removing only the first and last arguments could miss important flags.
+        # Safer approach is to include all relevant arguments except for optimization/debug flags
+        # file_args = file_args[1:-4] if len(file_args) > 5 else file_args
+
         clang_cmd = [
             "clang-14", self.config.source_file,
             "-std=c++17", '-emit-ast', '-o', '-']
         clang_cmd.extend(file_args)
 
-        clang_cmd_result = subprocess.run(clang_cmd, stdout=subprocess.PIPE)
-        clang_cmd_result.check_returncode()
+        try:
+            clang_cmd_result = subprocess.run(
+                clang_cmd, stdout=subprocess.PIPE)
+            clang_cmd_result.check_returncode()
+        except subprocess.CalledProcessError as e:
+            # Capture standard error and output
+            logging.error(
+                f"Clang failed with return code {e.returncode}. Error: {clang_cmd_result.stderr.decode()}")
+            raise RuntimeError(
+                f"Clang AST generation failed. See error log for details.")
 
         # Since `clang.Index.read` expects a file path, write generated abstract syntax tree to a
         # temporary named file. This file will be automatically deleted when closed.
         with tempfile.NamedTemporaryFile() as ast_file:
             ast_file.write(clang_cmd_result.stdout)
             self.ast = idx.read(ast_file.name)
 
+        self.assing_ast_with_ids(self.ast.cursor)
+
+        logging.info("AST generation completed successfully.")
+
+    def assing_ast_with_ids(self, cursor: Cursor) -> None:
+        """
+        Traverse the AST and assign a unique ID to each node during traversal.
+
+        Parameters:
+        @cursor: The current node (Cursor) in the AST to traverse.
+        """
+
+        self.node_counter += 1
+        cursor_id = self.node_counter
+
+        self.cursor_ids.append((cursor, cursor_id))
+        self.cursor_nodes.append(cursor)
+
+        logging.debug(
+            f"Node {cursor.spelling or cursor.kind} assigned ID {cursor_id}")
+
+        for child in cursor.get_children():
+            self.assing_ast_with_ids(child)
+
+    @staticmethod
+    def get_node_id(cursor: Cursor) -> int:
+        """
+        Returns the id of the current node.
+        """
+
+        for c, id in Scanner.cursor_ids:
+            if c == cursor:
+                return id
+        return 0
+
+    def get_node_by_index(self, index: int) -> Cursor:
+        """
+        Returns the node at the specified index position.
+        If the index is out of bounds, None is returned.
+        """
+        if 0 <= index < len(self.cursor_nodes):
```

Comment:
Instead of returning None we should raise an IndexError with a fitting error message